### DISCUSSION TITLE
Path parameters not set in starlette Request object

### DISCUSSION THREAD

**Discussion by shalseban** [fastapi/fastapi#7902] [Questions] [answered]
### Describe the bug
I am trying to access the `path_params` for a request from the middleware using `request.path_params` but it shows an empty `{}`.

However, it works if I access the `Request` object from within the endpoint code. 

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

```Python
from fastapi import FastAPI

app = FastAPI()
from starlette.requests import Request

@app.get('/fib-feed/{hello}/hellos')
async def test(path: str, query1: str, query2: str, request: Request):
    print(request.path_params)     # Is correctly populated


@app.middleware("http")
async def middleware(request: Request, call_next):
    print(request.path_params)     # Empty here
```

### Expected behavior

The `path_params` should be accessible in middleware too. 

### Environment

- OS: macOS
- FastAPI Version : 0.20.0 [e.g. 0.3.0], get it with:

- Python version, get it with:
 3.6.7

**Comment by nsidnev** [comment, marked as answer]
@shalseban 
Hi! I'm not sure if this should be a `FastAPI` issue. `FastAPI` does nothing with the middleware system provided by `Starlette`, so I think that this issue should be opened in their repository. 
The main problem here is that `path_params` in `Request` object are filled when the route is checked for matching with the request, and this happens after the middleware called `call_next` callable. This way you can get `path_params` in middleware, but only after your route has returned the response.

As a workaround you can manualy iterate over your application routes and get `path_params`, as the `Router` class from `Starlette` [does](https://github.com/encode/starlette/blob/5c43dde0ec0917673bb280bcd7ab0c37b78061b7/starlette/routing.py#L544) when matching request:
```python3
from typing import Callable

from fastapi import FastAPI
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Match

app = FastAPI()


@app.middleware("http")
async def my_middleware(request: Request, call_next: Callable) -> Response:
    routes = request.app.router.routes
    for route in routes:
        match, scope = route.matches(request)
        if match == Match.FULL:
            print(scope["path_params"])

    return await call_next(request)


@app.get("/{param}")
def my_route(param: str) -> dict:
    return {"param": param}
```

**Comment by shalseban** [comment]
@nsidnev 
Interesting. Thanks for clarifying that. I will close this and open in the starlette repo.

**Comment by inikolaev** [comment]
> @nsidnev
> Interesting. Thanks for clarifying that. I will close this and open in the starlette repo.

@shalseban have you opened an issue with Starlette regarding this? Could you reference it here if so?

**Comment by gjdanis** [comment]
Hey @shalseban I think this is still an issue. Any chance you could open this up in Starlette?

**Comment by tiangolo** [comment]
Thanks for the help here @nsidnev! 🙇 

And thanks for coming back to close the issue @shalseban. 🍪 

This is indeed not really related to FastAPI. And I think it wouldn't be easily doable in Starlette, maybe it wouldn't make sense fully, because the middleware has no idea about the routes that would handle the request, to extract the parameters. 🤔
